go/vt/vttablet/tabletmanager/vreplication/vcopier_copy_all.go

```diff
@@ -0,0 +1,297 @@
+
+	ctx, cancel := context.WithTimeout(ctx, copyPhaseDuration)
+	defer cancel()
+
+	rowsCopiedTicker := time.NewTicker(rowsCopiedUpdateInterval)
+	defer rowsCopiedTicker.Stop()
+	copyStateGCTicker := time.NewTicker(copyStateGCInterval)
+	defer copyStateGCTicker.Stop()
+
+	parallelism := getInsertParallelism()
+	copyWorkerFactory := vc.newCopyWorkerFactory(parallelism)
+	var copyWorkQueue *vcopierCopyWorkQueue
+
+	// Allocate a result channel to collect results from tasks.
+	resultCh := make(chan *vcopierCopyTaskResult, parallelism*4)
+	defer close(resultCh)
+
+	var lastpk *querypb.Row
+	var pkfields []*querypb.Field
+	var lastpkbv map[string]*querypb.BindVariable
+	// Use this for task sequencing.
+	var prevCh <-chan *vcopierCopyTaskResult
+	var gtid string
+
+	serr := vc.vr.sourceVStreamer.VStreamTables(ctx, func(resp *binlogdatapb.VStreamTablesResponse) error {
+		defer vc.vr.stats.PhaseTimings.Record("copy", time.Now())
+		defer vc.vr.stats.CopyLoopCount.Add(1)
+		log.Infof("VStreamTablesResponse: received table %s, #fields %d, #rows %d, gtid %s, lastpk %+v",
+			resp.TableName, len(resp.Fields), len(resp.Rows), resp.Gtid, resp.Lastpk)
+		tableName := resp.TableName
+		gtid = resp.Gtid
+
+		updateRowsCopied := func() error {
+			updateRowsQuery := binlogplayer.GenerateUpdateRowsCopied(vc.vr.id, vc.vr.stats.CopyRowCount.Get())
+			_, err := vc.vr.dbClient.Execute(updateRowsQuery)
+			return err
+		}
+
+		if err := updateRowsCopied(); err != nil {
+			return err
+		}
+		select {
+		case <-rowsCopiedTicker.C:
+			if err := updateRowsCopied(); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return io.EOF
+		default:
+		}
+		if tableName != state.currentTableName {
+			if copyWorkQueue != nil {
+				copyWorkQueue.close()
+			}
+			copyWorkQueue = vc.newCopyWorkQueue(parallelism, copyWorkerFactory)
+			if state.currentTableName != "" {
+				log.Infof("copy of table %s is done at lastpk %+v", state.currentTableName, lastpkbv)
+				if err := vc.deleteCopyState(state.currentTableName); err != nil {
+					return err
+				}
+			} else {
+				log.Infof("starting copy phase with table %s", tableName)
+			}
+
+			state.currentTableName = tableName
+		}
+		if !copyWorkQueue.isOpen {
+			if len(resp.Fields) == 0 {
+				return fmt.Errorf("expecting field event first, got: %v", resp)
+			}
+			fieldEvent := &binlogdatapb.FieldEvent{
+				TableName: tableName,
+			}
+			fieldEvent.Fields = append(fieldEvent.Fields, resp.Fields...)
+			tablePlan, err := state.plan.buildExecutionPlan(fieldEvent)
+			if err != nil {
+				return err
+			}
+			pkfields = append(pkfields, pkfields...)
+			buf := sqlparser.NewTrackedBuffer(nil)
+			buf.Myprintf(
+				"insert into _vt.copy_state (lastpk, vrepl_id, table_name) values (%a, %s, %s)", ":lastpk",
+				strconv.Itoa(int(vc.vr.id)),
+				encodeString(tableName))
+			addLatestCopyState := buf.ParsedQuery()
+			copyWorkQueue.open(addLatestCopyState, pkfields, tablePlan)
+		}
+		if len(resp.Rows) == 0 {
+			return nil
```

Comment:
Worth adding comment: this means that we have completed copying all of current table's rows (am I correct?)

Reply:
Added comment.